### ISSUE TITLE
assert priority > 0 (PER)

### ISSUE THREAD

**Issue by junhuang-ifast** [medipixel/rl_algorithms#191]
hi im using the Rainbow IQN and have run into [this](https://github.com/medipixel/rl_algorithms/blob/e20f61649dc904f9b7e647b5f47d3c82b2b38d29/rl_algorithms/common/buffer/priortized_replay_buffer.py#L143) assert error during training. 

From my analysis, it seems my [loss](https://github.com/medipixel/rl_algorithms/blob/e20f61649dc904f9b7e647b5f47d3c82b2b38d29/rl_algorithms/dqn/agent.py#L298) value has gone below -1e-6 since I set `per_eps=1e-6`. Note that I am using n-step so my loss (`dq_loss_element_wise`) is coming from [here](https://github.com/medipixel/rl_algorithms/blob/e20f61649dc904f9b7e647b5f47d3c82b2b38d29/rl_algorithms/dqn/agent.py#L271) . From [this](https://nbviewer.jupyter.org/github/Curt-Park/rainbow-is-all-you-need/blob/master/03.per.ipynb) tutorial, per_eps is to ensure all transitions are at least sampled once. I am thinking of doing this instead to prevent interruption during training:

```
# update priorities in PER
loss_for_prior = dq_loss_element_wise.detach().cpu().numpy()
new_priorities = torch.where(loss_for_prior>0, loss_for_prior+self.hyper_params.per_eps, torch.ones(loss_for_prior.shape)*self.hyper_params.per_eps)
self.memory.update_priorities(indices, new_priorities)
```

Do you think this change is an acceptable alternative and would not break anything?

However, I am concern that the fact that my loss would go to such large negative values, might mean that there is something fundamentally wrong with my network instead? (my network is a basic 1D-CNN). Any ideas one this issue?

**Comment by MrSyee** [comment]
Hi @junhuang-ifast 

What environment do you test? I've not seen the negative loss in environments that we tested like lunarlander, pong.

**Comment by junhuang-ifast** [comment]
@MrSyee im using my own custom environment (not gym related). Assuming there is nothing wrong with my environment, would you say this

```
# update priorities in PER
loss_for_prior = dq_loss_element_wise.detach().cpu().numpy()
torch.where(loss_for_prior>0, loss_for_prior, torch.ones(loss_for_prior.shape)*self.hyper_params.per_eps)
self.memory.update_priorities(indices, new_priorities)
```

method is an acceptable alternative even on gym environments?

**Comment by MrSyee** [comment]
@junhuang-ifast 
I wonder `dq_loss_element_wise` has the negative value because of [this line](https://github.com/medipixel/rl_algorithms/blob/e20f61649dc904f9b7e647b5f47d3c82b2b38d29/rl_algorithms/dqn/agent.py#L300). `per_eps` is the positive value and it adds up the loss value, so I suspect the loss value is negative value.
I also wonder the loss value that is got by [this code](https://github.com/medipixel/rl_algorithms/blob/e20f61649dc904f9b7e647b5f47d3c82b2b38d29/rl_algorithms/dqn/utils.py#L116) is negative value:
```
    quantile_huber_loss = (
        torch.abs(quantiles - (bellman_errors < 0).float().detach())
        * huber_loss
        / kappa
    )
```
I think this loss is always positive value. For this reason, this assert in PER code is reasonable.

When I tested on gym environment which this repo has by example, I've not seen the negative loss. So I can't reproduce this error without your setting. I think you should check your environment, config and code of making negative loss.

**Comment by junhuang-ifast** [comment]
I think the issue was with my environment. Thanks.
